web/components/auth-context.tsx

```diff
@@ -0,0 +1,71 @@
+import { createContext, useEffect } from 'react'
+import { User } from 'common/user'
+import { onIdTokenChanged } from 'firebase/auth'
+import {
+  auth,
+  listenForUser,
+  getUser,
+  setCachedReferralInfoForUser,
+} from 'web/lib/firebase/users'
+import { deleteAuthCookies, setAuthCookies } from 'web/lib/firebase/auth'
+import { createUser } from 'web/lib/firebase/api'
+import { randomString } from 'common/util/random'
+import { identifyUser, setUserProperty } from 'web/lib/service/analytics'
+import { useStateCheckEquality } from 'web/hooks/use-state-check-equality'
+
+const CACHED_USER_KEY = 'CACHED_USER_KEY'
+
+const ensureDeviceToken = () => {
+  let deviceToken = localStorage.getItem('device-token')
+  if (!deviceToken) {
+    deviceToken = randomString()
+    localStorage.setItem('device-token', deviceToken)
+  }
+  return deviceToken
+}
+
+export const AuthContext = createContext<User | null>(null)
+
+export function AuthProvider({ children }: any) {
+  const [currentUser, setCurrentUser] = useStateCheckEquality<User | null>(null)
```

Comment:
Concerned about the first render on the server and client. Previously, currentUser was undefined for this, then loaded from local storage on the first useEffect.

I think we should replicate that by setting the initial value to undefined. Some code handles undefined differently from null. Undefined means we're waiting to see whether you're logged in or not.

Reply:
OK, I agree that it seems reasonable to draw the distinction between "not sure yet" and "not logged in" for now. (I'm not sure it will still be reasonable in the future if we're doing SSR everywhere, but we can revisit it at a later date.)